### PR TITLE
Blog post about how to optimize LLMs for memory and speed

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This blog post does a deep dive into:

- Optimizing memory consumption with 8bit/4bit
- Optimizing speed with Flash Attention
- Explain smart MQA/GQA as well as Alibi/RoPE

# TODOs:
- [ ] Title is currently called "Getting the most out of LLMS" - is there a better title maybe? Maybe something like "Optimizing LLMs for memory and speed"? (wdyt @gante ?)
- [ ] Currently there is no nice thumbnail. Would be good to make one.

The corresponding `transformers` doc PR is here: https://github.com/huggingface/transformers/pull/26058

### CODE REVIEW

File: getting_the_most_out_of_LLMs.md [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+```
+
+For comparison, let's run the same function, but enable Flash Attention instead.
+To do so, we convert the model to [BetterTransformers](https://huggingface.co/docs/optimum/bettertransformer/overview) and by doing so enabling PyTorch's [SDPA self-attention](https://pytorch.org/docs/master/generated/torch.nn.functional.scaled_dot_product_attention) which in turn is based on Flash Attention.
+
+```python
+model.to_bettertransformer()
+```
+
+Now we run the exact same code snippet as before and under the hood Transformers will make use of Flash Attention.
```
Comment (younesbelkada) [inline getting_the_most_out_of_LLMs.md:443]: Note that to force-dispatch the SDPA kernels to use FA-1 you need to add a special context manager 

```diff
+ with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    result = pipe(**inputs)
```
As SDPA might dispatch the inference kernels to the math or mem_efficient kernels, in your example it is probably fine as the model is in half precision and you are using a GPU but worth highlighting to users that SDPA comes with three kernels, and if you specifically want to use FA you should use that context manager. 

You can find more info here: https://huggingface.co/docs/transformers/perf_infer_gpu_one#decoder-models

**Review by younesbelkada — COMMENTED** [review]
Very nice and great blogpost @patrickvonplaten ! 
I left few minor comments and open discussions, let me know what do you think!

Note that you can combine optimization tricks together, for example you can combine 8-bit / 4-bit and flash attention, I feel this is not clear to users and is definitely worth emphasizing it, what do you think?

File: getting_the_most_out_of_LLMs.md [hunk 2]
```diff
@@ -0,0 +1,734 @@
...
+
+As of writing this document, the largest GPU chip on the market is the A100 offering 80GB of VRAM. Most of the models listed before require more than 80GB just to be loaded and therefore necessarily require [tensor parallelism](https://huggingface.co/docs/transformers/v4.15.0/parallelism#tensor-parallelism) and/or [pipeline parallelism](https://huggingface.co/docs/transformers/v4.15.0/parallelism#naive-model-parallel-vertical-and-pipeline-parallel).
+
+🤗 Transformers does not support tensor parallelism out of the box as it requires the model architecture to be written in a specific way. If you're interested in writing models in a tensor-parallelism-friendly way, feel free to have a look at [the text-generation-inference library](https://github.com/huggingface/text-generation-inference/tree/main/server/text_generation_server/models/custom_modeling).
+
+Pipeline parallelism is supported out of the box. For this, simply load the model with `device="auto"` which will automatically place the different layers on the available GPUs as explained [here](https://huggingface.co/docs/accelerate/v0.22.0/en/concept_guides/big_model_inference).
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:64]: yeah good point! Should be more specific here!

File: getting_the_most_out_of_LLMs.md [hunk 3]
```diff
@@ -0,0 +1,734 @@
...
+
+If you have access to an 8 x 80GB A100 node, you could load BLOOM as follows
+
+```bash
+!pip install transformers accelerate bitsandbytes optimum
+```
+```python
+# from transformers import AutoModelForCausalLM
+
+# model = AutoModelForCausalLM.from_pretrained("bigscience/bloom", device_map="auto", pad_token_id=0)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:74]: I want to leave this as is actually since trying to download bloom will break the notebook (it's too big)

File: getting_the_most_out_of_LLMs.md [hunk 4]
```diff
@@ -0,0 +1,734 @@
...
+```
+9.543574333190918
+```
+
+Just 9.5GB! That's really not a lot for a >15 billion parameter model.
+
+While we see very little degradation in accuracy for our model here, 4-bit quantization can in practice often lead to different results compared to 8-bit quantization or full `bfloat16` inference. It is up to the user to try it out.
+
+Also note that inference here was again a bit slower compared to 8-bit quantization which is due to the more aggressive quantization method used for 4-bit quantization leading to $\text{quantize}$ and $\text{dequantize}$ taking longer during inference.
+
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:260]: Can I also run this in bfloat16? I'm running everything in bfloat16 here not in float16

File: getting_the_most_out_of_LLMs.md [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+Large Language Models (LLMs) such as GPT3/4, [Falcon](https://huggingface.co/tiiuae/falcon-40b), and [LLama](https://huggingface.co/meta-llama/Llama-2-70b-hf) are rapidly advancing in their ability to tackle human-centric tasks, establishing themselves as essential tools in modern knowledge-based industries.
+Deploying these models in real-world tasks remains challenging, however:
+
+-   To exhibit near-human text understanding and generation capabilities, LLMs currently require to be composed of billions of parameters (see [Kaplan et al](https://arxiv.org/abs/2001.08361), [Wei et. al](https://arxiv.org/abs/2206.07682)). This consequently amplifies the memory demands for inference.
+-   In many real-world tasks, LLMs need to be given extensive contextual information. This necessitates the model's capability to manage very long input sequences during inference.
+
+The crux of these challenges lies in augmenting the computational and memory capabilities of LLMs, especially when handling expansive input sequences.
+
+In this blog post, we will go over the most effective techniques to tackle these challenges for efficient LLM deployment:
```
Comment (patrickvonplaten) [inline getting_the_most_out_of_LLMs.md:27]:
```suggestion
In this blog post, we will go over the most effective techniques at the time of writing this blog post to tackle these challenges for efficient LLM deployment:
```

File: getting_the_most_out_of_LLMs.md [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
+ 
+```python
+flush()
+```
+
+Now what if your GPU does not have 32 GB of VRAM? It has been found that model weights can be quantized to 8-bit or 4-bits without a significant loss in performance (see [Dettmers et al.](https://arxiv.org/abs/2208.07339)).
+Model can be quantized to even 3 or 2 bits with an acceptable loss in performance as shown in the recent [GPTQ paper](https://huggingface.co/docs/transformers/main_classes/quantization#general-usage) 🤯.
+
+Without going into too many details, quantization schemes aim at reducing the precision of weights while trying to keep the model's inference results as accurate as possible (*a.k.a* as close as possible to bfloat16).
+Note that quantization works especially well for text generation since all we care about is choosing the *most likely next token* and don't really care about the exact values of the next token *logit* distribution.
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:155]: Don't 100% agree, think many models just run in argmax, but ok!

File: getting_the_most_out_of_LLMs.md [hunk 7]
```diff
@@ -0,0 +1,734 @@
...
+Without going into too many details, quantization schemes aim at reducing the precision of weights while trying to keep the model's inference results as accurate as possible (*a.k.a* as close as possible to bfloat16).
+Note that quantization works especially well for text generation since all we care about is choosing the *most likely next token* and don't really care about the exact values of the next token *logit* distribution. 
+All that matters is that the next token *logit* distribution stays roughly the same so that an `argmax` or `topk` operation gives the same results.
+
+There are various quantization techniques, which we won't discuss in detail here, but in general, all quantization techniques work as follows:
+
+-   1.  Quantize all weights to the target precision
+-   2.  Load the quantized weights, and pass the input sequence of vectors in bfloat16 precision
+-   3.  Dynamically dequantize weights to bfloat16 to perform the computation with their input vectors in bfloat16 precision
+-   4.  Quantize the weights again to the target precision after computation with their inputs.
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:163]: Maybe @younesbelkada here, but I think it does if we would leave the weights "dequantized" in fp16 precision they would blow up the memory no?

File: getting_the_most_out_of_LLMs.md [hunk 8]
```diff
@@ -0,0 +1,734 @@
...
+flush()
+```
+
+Overall, we saw that running OctoCoder in 8-bit precision reduced the required GPU VRAM from 32G GPU VRAM to only 15GB and running the model in 4-bit precision further reduces the required GPU VRAM to just a bit over 9GB.
+ 
+4-bit quantization allows the model to be run on GPUs such as RTX3090, V100, and T4 which are quite accessible for most people.
+
+For more information on quantization and to see how one can quantize models to require even less GPU VRAM memory than 4-bit, we recommend looking into the [`AutoGPTQ`](https://huggingface.co/docs/transformers/main/en/main_classes/quantization#autogptq-integration%60) implementation.
+
+> As a conclusion, it is important to remember that model quantization trades improved memory efficiency against accuracy and in most cases inference time.
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:275]: Hmm but only when the fp16 is memory bottlenecked no? :sweat_smile: But also ok!

File: getting_the_most_out_of_LLMs.md [hunk 9]
```diff
@@ -0,0 +1,734 @@
...
+
+```python
+model.to_bettertransformer()
+```
+
+Now we run the exact same code snippet as before and under the hood Transformers will make use of Flash Attention.
+
+```py
+start_time = time.time()
+result = pipe(long_prompt, max_new_tokens=60)[0]["generated_text"][len(long_prompt):]
```
Comment (patrickvonplaten) [inline getting_the_most_out_of_LLMs.md:447]:
```suggestion
with torch.backends.cuda.sdp_kernel(enable_flash=True, enable_math=False, enable_mem_efficient=False):
    result = pipe(long_prompt, max_new_tokens=60)[0]["generated_text"][len(long_prompt):]
```

File: getting_the_most_out_of_LLMs.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply getting_the_most_out_of_LLMs.md:265]: Yes you can also run it in bfloat16 but it will be slower AFAIK

File: getting_the_most_out_of_LLMs.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (younesbelkada) [reply getting_the_most_out_of_LLMs.md:168]: Yes, the dynamic de-quantization happens indeed for 4bit quantization. i.e. all layers are initially quantized in 4bit, then at inference they are dynamically de-quantized to perform the matmul in 16bit, then quantized back in 4bit

Screenshot 2023-09-15 at 09 46 00

For LLM.int8 (8bit) the weights are all quantized in 8-bit, and during inference only the weights that will be responsible of the 16-bit matmul will be de-quantized in 16-bit

File: getting_the_most_out_of_LLMs.md [hunk 10]
```diff
@@ -0,0 +1,734 @@
...
+
+**Output**:
+```bash
+29.0260648727417
+```
+ 
+Close enough to our back-of-the-envelope computation! We can see the number is not exactly correct as going from bytes to kilobytes requires a multiplication of 1024 instead of 1000. Therefore the back-of-the-envelope formula can also be understood as an "at most X GB" computation.
+Note that if we had tried to run the model in full float32 precision, a whopping 64 GB of VRAM would have been required.
+
+> Almost all models are trained in bfloat16 nowadays, there is no reason to run the model in full float32 precision if [your GPU supports bfloat16](https://discuss.pytorch.org/t/bfloat16-native-support/117155/5). Float32 won't give better inference results than the precision that was used to train the model.
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:128]: Good idea!

File: getting_the_most_out_of_LLMs.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply getting_the_most_out_of_LLMs.md:265]: Hmm ok maybe a bit too much for this blog post then. I'll leave as is for now - the user can check the docs for more info as mentioned below.

File: _blog.yml [hunk 11]
```diff
@@ -2777,7 +2777,7 @@
     - collaboration
     - diffusers
     - diffusion
-    
+
```
Comment (patrickvonplaten) [inline _blog.yml:2780]:
```suggestion
```

**Comment by patrickvonplaten** [comment]
Merging in invisible mode for now

**Comment by sayakpaul** [comment]
Let's please move the assets out of the directory.

**Comment by patrickvonplaten** [comment]
> Let's please move the assets out of the directory.

I don't understand this

**Comment by sayakpaul** [comment]
We only keep the thumbnails in the repository these days. This PR introduced additional assets (images) that are non-thumbnail ones. Those should reside in https://huggingface.co/datasets/huggingface/documentation-images.
